### ISSUE TITLE
Flat Interpolation patches no longer apply to Wine 2.8 code

### ISSUE THREAD

**Issue by gamax92** [gamax92/wine-overwatch#8]
There are three patches from the original wine-overwatch repository that due to changes in 2.8 no longer apply properly. All my attempts to rewrite them for 2.8 have resulted in glitchy black pixels or a black screen with massive GL errors being spammed to the console.

The current overwatch-2.8 branch does not have the patches applied.

These three original patches are:

https://bitbucket.org/awesie/wine-overwatch/diff/dlls/wined3d/glsl_shader.c?diff2=7cb22a538540&at=overwatch
https://bitbucket.org/awesie/wine-overwatch/diff/dlls/wined3d/glsl_shader.c?diff2=7adf9848895b&at=overwatch
https://bitbucket.org/awesie/wine-overwatch/diff/dlls/wined3d/glsl_shader.c?diff2=2f6d749ddf34&at=overwatch

And the conflicting wine commit is:

https://source.winehq.org/git/wine.git/commit/eb30543b1a55df61c0745e851fd2753a6227d188

@awesie If possible I'd like your input on this.

**Comment by gamax92** [comment]
I've gotten to the point where text is rendering and the world isn't a splotchy mess, but several reflective surfaces now flicker heavily. I've confirmed that without the patch, these surfaces do not flicker, so there isn't any regression in wine 2.8 or staging itself.

A good testing reference is the Oasis Gardens map, where the entire ceiling is a reflective surface.

My progress in this issue can be seen in the [flat-interpolation branch](https://github.com/gamax92/wine-overwatch/tree/flat-interpolation)

**Comment by awesie** [comment]
Here (https://gist.github.com/awesie/8f12c11e278f05073a8906a9d1b4f23b) is a patch that should fix the flat interpolation support. It is against your overwatch-2.8 branch.

I tested that it works with Overwatch without any obvious graphical glitches. I haven't tested to make sure it doesn't break anything else though.

Something to keep in mind with this code is that we want the output of a vertex shader to be an array if the next shader is a geometry shader; otherwise, the next shader is a pixel shader, and we don't want it to be an array (so we can do the interpolation stuff). And the corollary, we want the geometry shader input to be an array, and the pixel shader input to not be an array. Obviously, the output of the geometry shader should never be an array.

Hopefully I did that correctly in this patch. I haven't run the wine tests at all though.
